src/Parser/Deathknight/Unholy/Modules/Features/Apocalypse.js

```diff
@@ -0,0 +1,92 @@
+import React from 'react';
+
+import SPELLS from 'common/SPELLS';
+import SpellIcon from 'common/SpellIcon';
+import SpellLink from 'common/SpellLink';
+import { formatPercentage } from 'common/format';
+
+import Analyzer from 'Parser/Core/Analyzer';
+import Enemies from 'Parser/Core/Modules/Enemies';
+import StatisticBox, { STATISTIC_ORDER } from 'Main/StatisticBox';
+import { encodeTargetString } from 'Parser/Core/Modules/EnemyInstances';
+
+class Apocalypse extends Analyzer {
+  static dependencies = {
+    enemies: Enemies,
+  };
+
+  // used to track how many stacks a target has
+  targets = {};
+
+  totalApocalypseCasts = 0;
+  apocalypseWoundsPopped = 0;
+  
+  //These three sections are for keeping track of the amount of Festering Wounds on the target
+  on_byPlayer_applydebuffstack(event){
+    const spellId = event.ability.guid;
+    if(spellId === SPELLS.FESTERING_WOUND.id){
+		this.targets[encodeTargetString(event.targetID, event.targetInstance)] = event.stack;
+	}
+  }
+
+  on_byPlayer_removedebuffstack(event){
+    const spellId = event.ability.guid;
+    if(spellId === SPELLS.FESTERING_WOUND.id){
+		this.targets[encodeTargetString(event.targetID, event.targetInstance)] = event.stack;
+    }
+  }
+
+  on_byPlayer_removedebuff(event){
+    const spellId = event.ability.guid;
+    if(spellId === SPELLS.FESTERING_WOUND.id){
+		this.targets[encodeTargetString(event.targetID, event.targetInstance)] = 0;
+	}
+  }
+
+  //Logic that both counts the amount of Apocalypse cast by the player, as well as the amount of wounds popped by those apocalypse.
+  on_byPlayer_cast(event){
+    const spellId = event.ability.guid;
+    if(spellId === SPELLS.APOCALYPSE.id){
+		this.totalApocalypseCasts+=1;
+		if(this.targets.hasOwnProperty(encodeTargetString(event.targetID, event.targetInstance))) {
+			const currentTargetWounds = this.targets[encodeTargetString(event.targetID, event.targetInstance)];
```

Comment:
While I believe your method of buff tracking does work, it is already implemented between the Entity and EnemyInstances classes, making your logic in applydebuffstack, removedebuffstack, and removedebuff methods unneeded. No need to 'reinvent the wheel'. I'm sorry I didn't warn you about this sooner in Discord, I think I misunderstood what you were trying to do.

Instead of using the above lines to get buff stacks, you can do something like:
```
const target = enemies.getEntity(event); // can be null for a "not noteworthy" target
const currentTargetWounds = target && target.hasBuff(SPELLS.FESTERING_WOUND.id)
    ? target.getBuff(SPELLS.FESTERING_WOUND.id).stacks
    : 0;
```